node/src/reactor/joiner.rs

```diff
@@ -1024,13 +872,15 @@ impl reactor::Reactor for Reactor {
                 rng,
                 JoinerEvent::BlockHeaderByHeightFetcher(request.into()),
             ),
+            JoinerEvent::FinishedJoining { block_header } => {
+                self.linear_chain_sync = LinearChainSyncState::Done(block_header);
+                Effects::new()
+            }
         }
     }
 
     fn maybe_exit(&self) -> Option<ReactorExit> {
-        if self.linear_chain_sync.stopped_for_upgrade() {
```

Comment:
This is no longer possible, since I do not expose this state.

Reply:
Is it true that the new code won't be able to synchronize across upgrades? It will either sync in the era running previous version and then upgrade via the `LinearChain`-raised event or not sync at all if other validators have already moved on and there will be no one to respond to protocol state requests.

Reply:
> Is it true that the new code won't be able to synchronize across upgrades?

So right now this isn't implemented but there's a ticket #1966 

In general, you won't always be able to sync across upgrades.  There could be a non-backward-compatible change to the wire format or networking protocol for instance.